search/doc_vec.hpp

```diff
@@ -0,0 +1,322 @@
+#pragma once
+
+#include "base/assert.hpp"
+#include "base/string_utils.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include <boost/optional.hpp>
+
+namespace search
+{
+template <typename Token>
+struct TokenWeightPair
+{
+  TokenWeightPair() = default;
+
+  template <typename T>
+  TokenWeightPair(T && token, double weight) : m_token(std::forward<T>(token)), m_weight(weight)
+  {
+  }
+
+  bool operator<(TokenWeightPair const & rhs) const
+  {
+    if (m_token != rhs.m_token)
+      return m_token < rhs.m_token;
+    return m_weight < rhs.m_weight;
+  }
+
+  void Swap(TokenWeightPair & rhs)
+  {
+    m_token.swap(rhs.m_token);
+    std::swap(m_weight, rhs.m_weight);
+  }
+
+  // Returns squared weight of the token-weight pair.
+  double SqrWeight() const { return m_weight * m_weight; }
+
+  Token m_token;
+  double m_weight = 0;
+};
+
+template <typename Token>
+std::string DebugPrint(TokenWeightPair<Token> const & tw)
+{
+  std::ostringstream os;
+  os << "TokenWeightPair [ " << DebugPrint(tw.m_token) << ", " << tw.m_weight << " ]";
+  return os.str();
+}
+
+namespace impl
+{
+// Accumulates weights of equal tokens in |tws|. Result is sorted by
+// tokens. Also, maximum weight from a group of equal tokens will be
+// stored in the corresponding |maxWeight| elem.
+template <typename Token>
+void SortAndMerge(std::vector<TokenWeightPair<Token>> & tws, std::vector<double> & maxWeights)
+{
+  std::sort(tws.begin(), tws.end());
+  size_t n = 0;
+  maxWeights.clear();
+  for (size_t i = 0; i < tws.size(); ++i)
+  {
+    ASSERT_LESS_OR_EQUAL(n, i, ());
+    ASSERT_EQUAL(n, maxWeights.size(), ());
+
+    if (n == 0 || tws[n - 1].m_token != tws[i].m_token)
+    {
+      tws[n].Swap(tws[i]);
+      maxWeights.push_back(tws[n].m_weight);
+      ++n;
+    }
+    else
+    {
+      tws[n - 1].m_weight += tws[i].m_weight;
+      maxWeights[n - 1] = std::max(maxWeights[n - 1], tws[i].m_weight);
+    }
+  }
+
+  ASSERT_LESS_OR_EQUAL(n, tws.size(), ());
+  tws.erase(tws.begin() + n, tws.end());
+}
+
+// Computes squared L2 norm of vector of tokens.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws)
+{
+  double sum = 0;
+  for (auto const & tw : tws)
+    sum += tw.SqrWeight();
+  return sum;
+}
+
+// Computes squared L2 norm of vector of tokens + prefix token.
+template <typename Token>
+double SqrL2(std::vector<TokenWeightPair<Token>> const & tws,
+             boost::optional<TokenWeightPair<Token>> const & prefix)
+{
+  double result = SqrL2(tws);
+  return result + (prefix ? prefix->SqrWeight() : 0);
+}
+}  // namespace impl
+
+// This class represents a document in a vector space of tokens.
+template <typename Token>
+class DocVec
```

Comment:
Are you sure this name is good enough? Why not DocumentVector?

Reply:
Yes, I'm sure. Can't find any reason to switch from Doc to Document now.